core/bloombits/generator.go

```diff
@@ -65,18 +65,21 @@ func (b *Generator) AddBloom(index uint, bloom types.Bloom) error {
 	}
 	// Rotate the bloom and insert into our collection
 	byteIndex := b.nextSec / 8
-	bitMask := byte(1) << byte(7-b.nextSec%8)
-
-	for i := 0; i < types.BloomBitLength; i++ {
-		bloomByteIndex := types.BloomByteLength - 1 - i/8
-		bloomBitMask := byte(1) << byte(i%8)
-
-		if (bloom[bloomByteIndex] & bloomBitMask) != 0 {
-			b.blooms[i][byteIndex] |= bitMask
-		}
+	bitIndex := byte(7 - b.nextSec%8)
+	for byt := 0; byt < types.BloomByteLength; byt++ {
+		bloomByteIndex := types.BloomByteLength - 1 - byt
+		bloomByte := bloom[bloomByteIndex]
+		base := 8 * byt
+		b.blooms[base+7][byteIndex] |= (bloomByte & 0x80 >> 7) << bitIndex
```

Comment:
I'm not 100% sure, but I think >> has a precedence over & which would mean you could precompute (0x80 >> 7) which is 1. So either you need brackets around (0x80>>7) or you could just do bloomByte & 1 in every clause.

Reply:
Precedence table for go: https://www.tutorialspoint.com/go/go_operators_precedence.htm#:~:text=Operator%20precedence%20determines%20the%20grouping,how%20an%20expression%20is%20evaluated.&text=For%20example%20x%20%3D%207%20%2B%203,and%20then%20adds%20into%207.

Reply:
https://play.golang.org/p/RZH5cI3trRG

Reply:
The thing that we're doing here, is picking out each individual bit, and then OR:ing it into a 'column'. So essentially mapping out vertical bits horizontally. And that would insta-fail if I had gotten the operators wrong.

Reply:
Ahh I see, they have the same precedence which means they evaluate from left to right

Reply:
You have a point though. If I rightshift first, then the code becomes simpler (because every line will have identical 0x1 bitmasking). I'll change that